src/backend/columnar/columnar_reader.c

```diff
@@ -1042,16 +1042,44 @@ GetClauseVars(List *whereClauseList, int natts)
 	 */
 	int flags = 0;
 	List *vars = pull_var_clause((Node *) whereClauseList, flags);
-	Var **deduplicate = palloc0(sizeof(Var *) * natts);
+	return DeduplicateVarList(vars, natts);
+}
+
+
+/*
+ * DeduplicateVarList takes a list of Var objects, and deduplicates and
+ * sorts them in a new list.
+ *
+ * Note that this function doesn't expect any Var's referencing to a system
+ * column or to an invalid attribute. If given varList contains such a Var,
+ * then throws an error.
+ */
+List *
+DeduplicateVarList(List *varList, int natts)
+{
+	Var **deduplicate = palloc0(sizeof(Var *) * (natts));
 
 	ListCell *lc;
-	foreach(lc, vars)
+	foreach(lc, varList)
 	{
 		Node *node = lfirst(lc);
 		Assert(IsA(node, Var));
 
 		Var *var = (Var *) node;
+		if (var->varattno == 0)
+		{
+			/* whole-row reference dominates all */
+			pfree(deduplicate);
+			return list_make1(var);
+		}
```

Comment:
Codecov complains that we don't test whole-row references, not sure why below test doesn't trigger that code-path:

```sql
EXPLAIN (COSTS OFF, SUMMARY OFF)
SELECT *, "bbbbbbbbbbbbbbbbbbbbbbbbb\!bbbb'bbbbbbbbbbbbbbbbbbbbb''bbbbbbbb"
FROM weird_col_explain
WHERE "bbbbbbbbbbbbbbbbbbbbbbbbb\!bbbb'bbbbbbbbbbbbbbbbbbbbb''bbbbbbbb" * 2 >
      "aaaaaaaaaaaa$aaaaaa$$aaaaaaaaaaaaaaaaaaaaaaaaaaaaa'aaaaaaaa'$a'!";
```

Reply:
`SELECT weird_col_explain FROM weird_col_explain`